gateway/reverse_proxy.go

```diff
@@ -785,73 +790,126 @@ func (rt *TykRoundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
 }
 
 const (
-	inMemNetworkName = "in-mem-network"
-	inMemNetworkType = "memu"
+	checkIdleMemConnInterval = 5 * time.Minute
+	maxIdleMemConnDuration   = -time.Minute
+	inMemNetworkName         = "in-mem-network"
+	inMemNetworkType         = "memu"
 )
 
-var (
-	memconnProvidersMtx sync.Mutex
-	// TODO: Find a proper way to clean that cache
-	memconnProvidersCache = make(map[string]*memconn.Provider)
-)
+type memConnProvider struct {
+	listener net.Listener
+	provider *memconn.Provider
+	lastUsed int64
+}
 
-func createMemconnProviderIfNeeded(handler http.Handler, r *http.Request) error {
-	memconnProvidersMtx.Lock()
-	defer memconnProvidersMtx.Unlock()
-	_, ok := memconnProvidersCache[r.Host]
+var memConnProviders = &struct {
```

Comment:
Is using &struct some particular convention that's being followed? Anonymous structs don't have function receivers, and pointer semantics don't make any sense for globals